File: backend/app/routers/groups.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from ..database import get_db
from ..schemas.group import (
    GroupCreate,
    GroupDetail,
    GroupMatchResponse,
    GroupMessageCreate,
    GroupMessagePage,
    GroupMessageRead,
    GroupRead,
    InviteLinkResponse,
    JoinGroupRequest,
)
from ..schemas.scheduling import (
    AvailabilityCreate,
    AvailabilityRead,
    GroupMeetingRead,
    MeetingConfirmationRequest,
    MeetingPreferences,
    MeetingSuggestionResponse,
)
from ..services.groups import GroupQueryService, GroupService
from ..services.matching import MatchingService
from ..services.scheduling import AvailabilityService, SchedulingService

router = APIRouter(prefix="/groups", tags=["groups"])


@router.post("/", response_model=GroupRead, status_code=status.HTTP_201_CREATED)
def create_group(payload: GroupCreate, db: Session = Depends(get_db)) -> GroupRead:
    group = GroupService.create_group(
        db,
        name=payload.name,
        description=payload.description,
        owner_id=payload.owner_id,
    )
    db.commit()
    return group


@router.get("/", response_model=list[GroupRead])
def list_groups(db: Session = Depends(get_db)) -> list[GroupRead]:
    groups = GroupQueryService.list_groups(db)
    return list(groups)


@router.get("/{group_id}", response_model=GroupDetail)
def retrieve_group(group_id: str, db: Session = Depends(get_db)) -> GroupDetail:
    group = GroupService.get_group(db, group_id)
    return group


@router.post("/{group_id}/join", response_model=GroupDetail)
def join_group(
    group_id: str,
    payload: JoinGroupRequest,
    db: Session = Depends(get_db),
) -> GroupDetail:
    group = GroupService.join_group(
        db,
        group_id=group_id,
        user_id=payload.user_id,
        invite_code=payload.invite_code,
    )
    db.commit()
    return GroupService.get_group(db, group.id)


@router.get("/{group_id}/invite", response_model=InviteLinkResponse)
def get_invite_link(group_id: str, db: Session = Depends(get_db)) -> InviteLinkResponse:
    group = GroupService.get_group(db, group_id)
    url = GroupService.get_invite_link(group)
    return InviteLinkResponse(group_id=group.id, invite_code=group.invite_code, shareable_url=url)


@router.post(
    "/{group_id}/messages",
    response_model=GroupMessageRead,
    status_code=status.HTTP_201_CREATED,
)
def post_message(
    group_id: str,
    payload: GroupMessageCreate,
    db: Session = Depends(get_db),
) -> GroupMessageRead:
    message = GroupService.post_message(
        db,
        group_id=group_id,
        user_id=payload.user_id,
        content=payload.content,
    )
    db.commit()
    return message


@router.get("/{group_id}/messages", response_model=GroupMessagePage)
def list_messages(
    group_id: str,
    *,
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
) -> GroupMessagePage:
    page = GroupService.list_messages(db, group_id=group_id, limit=limit, offset=offset)
    return GroupMessagePage(messages=page.messages, total=page.total)


@router.post(
    "/{group_id}/availability",
    response_model=AvailabilityRead,
    status_code=status.HTTP_201_CREATED,
)
def add_availability(
    group_id: str,
    payload: AvailabilityCreate,
    db: Session = Depends(get_db),
) -> AvailabilityRead:
    record = AvailabilityService.add_window(
        db,
        group_id=group_id,
        user_id=payload.user_id,
        start_time=payload.start_time,
        end_time=payload.end_time,
        timezone_name=payload.timezone,
    )
    db.commit()
    return record


@router.get("/{group_id}/availability", response_model=list[AvailabilityRead])
def list_availability(group_id: str, db: Session = Depends(get_db)) -> list[AvailabilityRead]:
    return AvailabilityService.list_group_windows(db, group_id=group_id)


@router.post("/{group_id}/meeting-suggestions", response_model=MeetingSuggestionResponse)
@router.post(
    "/{group_id}/suggestions",
    response_model=MeetingSuggestionResponse,
    include_in_schema=False,
)
def suggest_meetings(
    group_id: str,
    preferences: MeetingPreferences | None = None,
    db: Session = Depends(get_db),
) -> MeetingSuggestionResponse:
    suggestions = SchedulingService.suggest_meetings(
        db,
        group_id=group_id,
        preferences=preferences,
    )
    return MeetingSuggestionResponse(suggestions=suggestions)


@router.post(
    "/{group_id}/meetings",
    response_model=GroupMeetingRead,
    status_code=status.HTTP_201_CREATED,
)
def confirm_meeting(
    group_id: str,
    payload: MeetingConfirmationRequest,
    db: Session = Depends(get_db),
) -> GroupMeetingRead:
    meeting = SchedulingService.confirm_meeting(
        db,
        group_id=group_id,
        scheduled_start=payload.start_time,
        scheduled_end=payload.end_time,
        suggested_by=payload.user_id,
        note=payload.title,
    )
    db.commit()
    return meeting


@router.get("/{group_id}/matches", response_model=GroupMatchResponse)
def generate_matches(
    group_id: str,
    limit: int = Query(5, ge=1, le=20),
    db: Session = Depends(get_db),
) -> GroupMatchResponse:
    candidates = MatchingService.generate_group_matches(db, group_id=group_id, limit=limit)
    return GroupMatchResponse(candidates=candidates)

File: frontend/src/services/groups.ts
import api from '../lib/api';

// Types matching backend schemas
export interface GroupCreate {
  name: string;
  description?: string;
  owner_id: string;
}

export interface GroupRead {
  id: string;
  name: string;
  description?: string;
  created_at: string;
}

export interface GroupMember {
  id: string;
  display_name?: string;
  email: string;
  role: 'owner' | 'member';
}

export interface GroupDetail extends GroupRead {
  members: GroupMember[];
}

export interface JoinGroupRequest {
  user_id: string;
  invite_code: string;
}

export interface InviteLinkResponse {
  invite_url: string;
}

export interface GroupMessageCreate {
  user_id: string;
  content: string;
}

export interface GroupMessageRead {
  id: string;
  group_id: string;
  user_id: string;
  content: string;
  created_at: string;
}

export interface GroupMessagePage {
  messages: GroupMessageRead[];
  total: number;
}

export interface AvailabilityCreate {
  user_id: string;
  start_time: string;
  end_time: string;
  timezone: string;
}

export interface AvailabilityRead extends AvailabilityCreate {
  id: string;
}

export interface MeetingPreferences {
  duration_minutes?: number;
  window_days?: number;
  limit?: number;
}

export interface MeetingSuggestion {
  start_time: string;
  end_time: string;
  participant_ids: string[];
  conflicts: string[];
}

export interface MeetingSuggestionResponse {
  suggestions: MeetingSuggestion[];
}

export interface MeetingConfirmationRequest {
  user_id: string;
  start_time: string;
  end_time: string;
  title?: string;
}

export interface GroupMeetingRead {
  id: string;
  group_id: string;
  scheduled_start: string;
  scheduled_end: string;
  title?: string;
  suggested_by?: string;
  created_at: string;
}

export interface GroupMatchCandidate {
  group_id: string;
  group_name: string;
  compatibility_score: number;
  overlap_minutes: number;
  size: number;
}

export interface GroupMatchResponse {
  candidates: GroupMatchCandidate[];
}

// API functions
export const groupsApi = {
  // Group CRUD
  createGroup: (data: GroupCreate) =>
    api.post<GroupRead>('/groups/', data),

  listGroups: () =>
    api.get<GroupRead[]>('/groups/'),

  getGroup: (groupId: string) =>
    api.get<GroupDetail>(`/groups/${groupId}`),

  // Join & Invite
  joinGroup: (groupId: string, data: JoinGroupRequest) =>
    api.post<GroupDetail>(`/groups/${groupId}/join`, data),

  getInviteLink: (groupId: string) =>
    api.get<InviteLinkResponse>(`/groups/${groupId}/invite`),

  // Messages
  postMessage: (groupId: string, data: GroupMessageCreate) =>
    api.post<GroupMessageRead>(`/groups/${groupId}/messages`, data),

  listMessages: (groupId: string, limit = 50, offset = 0) =>
    api.get<GroupMessagePage>(`/groups/${groupId}/messages`, {
      params: { limit, offset },
    }),

  // Availability
  addAvailability: (groupId: string, data: AvailabilityCreate) =>
    api.post<AvailabilityRead>(`/groups/${groupId}/availability`, data),

  listAvailability: (groupId: string) =>
    api.get<AvailabilityRead[]>(`/groups/${groupId}/availability`),

  // Meeting Suggestions
  getMeetingSuggestions: (groupId: string, preferences?: MeetingPreferences) =>
    api.post<MeetingSuggestionResponse>(
      `/groups/${groupId}/meeting-suggestions`,
      preferences || {}
    ),

  // Confirm Meeting
  confirmMeeting: (groupId: string, data: MeetingConfirmationRequest) =>
    api.post<GroupMeetingRead>(`/groups/${groupId}/meetings`, data),

  // Group Matches
  getGroupMatches: (groupId: string) =>
    api.get<GroupMatchResponse>(`/groups/${groupId}/matches`),
};
